### PR TITLE
 workDivMembers construction with a user defined type

### PR DESCRIPTION
**Pull request by mehmetyusufoglu**
`WorkDivMembers` class constructor could accept any vector type if the necessary traits and functions related to that type are provided.

This PR uses a user-defined vector, related traits and specializations to test the  of `WorkDivMembers` constructors. 

- Tests are added to test the usage of the user-defined vectors in `WorkDivMembers` construction.
-  Already existing `WorkDivMembers` construction tests are improved and refactored. 
-  Initializer list test is added in construction of `WorkDivMembers`.  `auto const workDiv2DUsingInitList = alpaka::WorkDivMembers<Dim2D, Idx>({6, 9}, {4, 6}, {1, 1});` initializes 3 alpaka vectors as arguments.

### CODE REVIEW

**Comment by fwyzard** [comment]
Rather than `std::array`s (or in addition to them, if there is any concrete use case), could you add support for `[std::initializer_list](https://en.cppreference.com/w/cpp/utility/initializer_list)` arguments ?

That would support `alpaka::WorkDivMembers{{1u, 1u}, {2u, 2u}, {1u, 1u}};`.

File: include/alpaka/vec/Vec.hpp [hunk 1]
```diff
@@ -656,6 +656,25 @@ namespace alpaka
         return r;
     }
 
+    //! \brief Converts std::array to alpaka Vec
+    //! \tparam TVal is the element type of the array
+    //! \tparam N is the size of the array
+    template<typename TVal, size_t N>
+    ALPAKA_FN_HOST_ACC constexpr auto arrayToVec(std::array<TVal, N> const& a) -> Vec<DimInt<N>, TVal>
```
Comment (psychocoderHPC) [inline include/alpaka/vec/Vec.hpp]: There is no need for an additional method that can only handle std array.
We have the traits `GetExtents` to provide the translation that alpaka can work with alpaka foreign types.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: include/alpaka/workdiv/WorkDivMembers.hpp [hunk 2]
```diff
@@ -33,6 +35,33 @@ namespace alpaka
...
+            , m_threadElemExtent(elemExtent)
+        {
+        }
+
+        //! \brief Accepts single specific type and is called without explicit template parameters.
+        //! \tparam T is the element type of the array
+        //! \tparam N is the size of the array
+        template<class T, size_t N>
+        ALPAKA_FN_HOST_ACC explicit WorkDivMembers(
+            std::array<T, N> const gridBlockExtent,
```
Comment (psychocoderHPC) [inline include/alpaka/workdiv/WorkDivMembers.hpp]: The goal should not be to add a constructor which understand `std::array`. We like to stay interpotable with any alpaka foreign type which represents extents. `std::array` is one example it could also be `std::vector` or users type like in PIConGPU `pmacc::math::Vector<>`. 
To reach this goal we have the traits these give the alpaka user the possibility to be specialize for there own types.

This allows alpaka to be very generic and therefore we do not need to write constructors for any typical used type. As sson as we have C++20 we could write everything on a much nicer why but until than we need to life with the trais.

**Comment by bernhardmgruber** [comment]
> That would support `alpaka::WorkDivMembers{{1u, 1u}, {2u, 2u}, {1u, 1u}};`.

This may already work with #2240, if the right constructor is selected. I think `alpaka::Vec` can be created from an initializer list.

**Comment by bernhardmgruber** [comment]
> > That would support `alpaka::WorkDivMembers{{1u, 1u}, {2u, 2u}, {1u, 1u}};`.
> 
> This may already work with #2240, if the right constructor is selected. I think `alpaka::Vec` can be created from an initializer list.

It turns out it doesn't work, and I am not exactly sure why. The compiler fails to deduce the template parameters of the deduction guide. It has something to do with `{1u, 1u}` being an initializer list, because `alpaka::WorkDivMembers{alpaka::Vec{1u, 1u}, {2u, 2u}, {1u, 1u}};` works successfully.

**Comment by mehmetyusufoglu** [comment]
Yes, i also tried. Did not work.

26 Şub 2024 Pzt 21:23 tarihinde Bernhard Manfred Gruber <
***@***.***> şunu yazdı:

File: include/alpaka/vec/Vec.hpp (hunk 1 above)
Comment (mehmetyusufoglu) [reply include/alpaka/vec/Vec.hpp]: Ok I used getExtents. There is already one implemented for 1D array arguments in "ViewPlanePtr" therefore I had to check the dimension. I will refine the solution. Thanks a lot.

File: include/alpaka/workdiv/WorkDivMembers.hpp [hunk 3]
```diff
@@ -33,6 +33,32 @@ namespace alpaka
...
 
+        //! Accepts different alpaka vector types and takes the last TDim number of items. TDim is explicitly specified
+        //! at the constructor call.
+        ALPAKA_NO_HOST_ACC_WARNING
+        template<typename TGridBlockExtent>
+        ALPAKA_FN_HOST_ACC explicit WorkDivMembers(
+            TGridBlockExtent const& gridBlockExtent = TGridBlockExtent(),
+            TGridBlockExtent const& blockThreadExtent = TGridBlockExtent(),
+            TGridBlockExtent const& threadElemExtent = TGridBlockExtent())
+            : m_gridBlockExtent(getExtents<TGridBlockExtent>(gridBlockExtent))
```
Comment (psychocoderHPC) [inline include/alpaka/workdiv/WorkDivMembers.hpp]: this should be `: m_gridBlockExtent(getExtents(gridBlockExtent))`

Comment (mehmetyusufoglu) [reply include/alpaka/workdiv/WorkDivMembers.hpp]: This part is left to another PR. For now `getExtents` is not called directly at the constructor.

File: include/alpaka/extent/Traits.hpp [hunk 4]
```diff
@@ -55,10 +55,10 @@ namespace alpaka
 
     //! \return The extents of the given object.
     ALPAKA_NO_HOST_ACC_WARNING
-    template<typename T>
-    ALPAKA_FN_HOST_ACC auto getExtents(T const& object) -> Vec<Dim<T>, Idx<T>>
+    template<typename TObject>
```
Comment (psychocoderHPC) [inline include/alpaka/extent/Traits.hpp]: changes from this file should not be part of this PR

File: test/unit/workDiv/CMakeLists.txt [hunk 5]
```diff
@@ -6,10 +6,13 @@
 set(_TARGET_NAME "workDivTest")
 
 append_recursive_files_add_to_src_group("src/" "src/" "cpp" _FILES_SOURCE)
+append_recursive_files_add_to_src_group("src/" "src/" "hpp" _FILES_HEADER)
```
Comment (psychocoderHPC) [inline test/unit/workDiv/CMakeLists.txt]: This should not be required because the file is directly included the cpp.

File: test/unit/workDiv/CMakeLists.txt [hunk 6]
```diff
@@ -6,10 +6,13 @@
...
 
 alpaka_add_executable(
     ${_TARGET_NAME}
-    ${_FILES_SOURCE})
+    ${_FILES_SOURCE}
+    ${_FILES_HEADER})
+
 target_link_libraries(
```
Comment (psychocoderHPC) [inline test/unit/workDiv/CMakeLists.txt:18]: Additionally 

```
target_include_directories(${_TARGET_NAME} PRIVATE "src")
```

is required to provide the target with the include directory where the hpps are located.

File: test/unit/workDiv/src/WorkDivHelpersTest.cpp [hunk 7]
```diff
@@ -34,6 +34,71 @@ namespace
     }
 } // namespace
 
+namespace alpaka::trait
```
Comment (psychocoderHPC) [inline test/unit/workDiv/src/WorkDivHelpersTest.cpp]: could you move this to the FooVec, please? No need to have this code here in the test file since it corresponds to the FooVec class.

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: test/unit/workDiv/CMakeLists.txt (hunk 6 above)
Comment (mehmetyusufoglu) [reply test/unit/workDiv/CMakeLists.txt:18]: Yes, sure. Actually,  I noticed a different usage in   https://github.com/alpaka-group/alpaka/blob/develop/test/unit/math/CMakeLists.txt  and used the same approach.

**Comment by mehmetyusufoglu** [comment]
> Rather than `std::array`s (or in addition to them, if there is any concrete use case), could you add support for `[std::initializer_list](https://en.cppreference.com/w/cpp/utility/initializer_list)` arguments ?
> 
> That would support `alpaka::WorkDivMembers{{1u, 1u}, {2u, 2u}, {1u, 1u}};`.

Currently the following are possible due to a merged PR:  `alpaka::WorkDivMembers<Dim2D,Idx>{{1u, 1u}, {2u, 2u}, {1u, 1u}};` .  OR  `alpaka::WorkDivMembers{Vec{1u, 1u}, Vec{2u, 2u}, Vec{1u, 1u}};`  Vec is alpaka::vec with specific Dim and Idx. 

I will consider your suggestion together with C++20, seems more reasonable I think.

File: test/unit/workDiv/src/FooVec.hpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+    template<typename T>
+    struct has_member_arr<T, std::void_t<decltype(&T::arr)>> : std::true_type
+    {
+    };
+
+    //! The DimType specialization for the user defined vector which has a member called "arr"
+    //! \tparam T The vector type
+    template<typename T>
+    struct DimType<T, std::enable_if_t<has_member_arr<T>::value>>
```
Comment (bernhardmgruber) [inline test/unit/workDiv/src/FooVec.hpp]:
```suggestion
    //! The DimType specialization for the user defined vector which has a member called "arr"
    //! \tparam T The vector type
    template<typename T>
    struct DimType<FooVec>
```
Why don't you just specialize for `FooVec` directly? I don't see a need for SFINAE here.

Applies to other traits as well.

File: test/unit/workDiv/src/FooVec.hpp [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+    // Move Assignment Operator
+    FooVec& operator=(FooVec&& other) noexcept
+    {
+        if(this != &other)
+        {
+            arr = std::move(other.arr);
+        }
+        return *this;
+    }
```
Comment (bernhardmgruber) [inline test/unit/workDiv/src/FooVec.hpp]:
```suggestion
```
I don't see a reason for implementing special member functions.

**Review by bernhardmgruber — COMMENTED** [review]
Here are a few comments:

File: test/unit/workDiv/src/FooVec.hpp (hunk 8 above)
Comment (mehmetyusufoglu) [reply test/unit/workDiv/src/FooVec.hpp]: FooVec is a template, not a type. But i will try to formulate your approach.

File: test/unit/workDiv/src/FooVec.hpp (hunk 9 above)
Comment (mehmetyusufoglu) [reply test/unit/workDiv/src/FooVec.hpp]: Ok removed. Might be misleading. Thanks.

File: test/unit/workDiv/src/FooVec.hpp (hunk 8 above)
Comment (bernhardmgruber) [reply test/unit/workDiv/src/FooVec.hpp]: Sorry, my bad. Should be:
```c++
template<typename TVec, std::size_t Size>
struct DimType<FooVec<TVec, Size>>
```

**Comment by mehmetyusufoglu** [comment]
Yes, i've already did like this actually. Testing,.

Bernhard Manfred Gruber ***@***.***>, 8 Mar 2024 Cum, 13:37
tarihinde şunu yazdı:

File: test/unit/workDiv/src/FooVec.hpp (hunk 8 above)
Comment (mehmetyusufoglu) [reply test/unit/workDiv/src/FooVec.hpp]: done thanks.

**Review by psychocoderHPC — APPROVED** [review]
